### ISSUE TITLE
Add support for S3-compatible object-lock properties

### ISSUE THREAD

**Issue by taupehat** [rclone/rclone#4683]
Please see [this forum thread](https://forum.rclone.org/t/requesting-support-for-s3-compatible-object-lock-properties/19675) for initial details including links to the API calls supported by both AWS and Backblaze.

Currently AWS and Backblaze (via their S3 API only) support object immutability. I am proposing a set of flags which would leverage this feature in supported object storage systems. A brief (and probably incomplete) overview of suggested flags follows:

### At bucket creation
` --s3-bucket-object-lock-enabled` - for "rclone mkdir" this enables object-locking for a bucket. It's important to note this _must_ be set at the time a bucket is created for objects inside that bucket to be lockable.

### At object creation/upload
`--s3-object-lock-mode` [GOVERNANCE|COMPLIANCE]
`--s3-object-lock-retain-until-date` [date]
`--s3-object-lock-retention-period` [days]
`--s3-object-lock-legal-hold` [ON|OFF] default off

### At object deletion
` --s3-bypass-governance-retention`

### Determine lock status
`rclone backend list-object-lock`

Additionally, the configurator for S3 would need to know about this feature and which APIs support it so a user would get big scary warnings if they tried to set object-locking in an incompatible provider. This probably means Backblaze should be added to the S3 configurator in addition to the current native B2 configuration.

**Comment by buchmann** [comment]
Will this support also data migration/copy between different Object Storages ? Example from OnPremise Objectstorage to Cloud Objecstorage and keep the Object Lock Retention Period from Source to Target ?

**Comment by UbuntuRunner** [comment]
Did I misunderstand something here or shouldn't the object lock feature be incorporated from version 1.54?

**Comment by ncw** [comment]
This hasn't been implemented yet - if anyone would like to help it would be much appreciated.

**Comment by monty-dev** [comment]
I wanted to provide some info that might help someone decide to take this on. I don't currently have the skill to contribute. 
Adding object lock appears to need 3 values: 

`ObjectLockLegalHoldStatus ` enum values `ON` `OFF`
`ObjectLockMode` enum values `GOVERNANCE` `COMPLIANCE`
`ObjectLockRetainUntilDate` time.Time 

It would be ideal if these values could be provided at config as a default, and also on the upload operation. 
sources:
[go sdk options](https://github.com/aws/aws-sdk-go/blob/main/service/s3/s3manager/upload_input.go#L124)
[s3 api values]( https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html)
[s3 object lock dev guide](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html)

Hope this is helpful.

**Comment by pmrt** [comment]
If anyone really needs this right now like me to prevent ransomware attacks, as a workaround you can manually provide the headers of your storage provider.

For backblaze, this works for me:
`rclone sync /path/to/dir <remote-name>:<bucket-name> --header-upload "X-Bz-File-Retention-Mode: governance" --header-upload "X-Bz-File-Retention-Retain-Until-Timestamp: 1637600400" --dry-run`

Also, this utility functions in Bash may be useful for your workaround:

```bash
# Get current date +180 days in epoch time in ms (actually nanoseconds truncated 
# to the 3 most significant digits) 
date -d "+180 days" +"%s%3N"
# 
```

**Comment by ncw** [comment]
This feature should be quite straight forward. If someone wants to have a go I'd be happy to mentor a PR.

**Comment by trompx** [comment]
Thanks for the workaround @pmrt.
I haven't tried it yet but for AWS, here are the doc to pass the retention mode & period:

- https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectRetention.html
- https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectLegalHold.html

**Comment by trompx** [comment]
I tried to copy files with s3 object lock (aws provider) with the command:
```sh
rclone --progress copy /src awss3crypt:/dest --header-upload "x-amz-object-lock-mode: GOVERNANCE" --header-upload "x-amz-object-lock-retain-until-date: ${RETENTION}"
```
but got the error:
```
2022-01-12 12:07:31 ERROR : XXX.bin: Don't know how to set key "x-amz-object-lock-mode" on upload
2022-01-12 12:07:31 ERROR : XXX.bin: Don't know how to set key "x-amz-object-lock-retain-until-date" on upload
```

Is there a way to pass headers that are not pre-determined here https://github.com/rclone/rclone/blob/master/backend/s3/s3.go#L3709 ?

**Comment by Panzer1119** [comment]
Could there be an option to extend Object Locks of already existing objects/files while a sync/copy happens?

So that it reads the Buckets default retention (for Object Lock) and applies it to all currently existing Files?

**Comment by prathamesh-gharat** [comment]
> Could there be an option to (automatically) extend Object Locks of already existing objects/files while a sync/copy happens?
> 
> So that it reads the Buckets default retention (for Object Lock) and applies it to all currently existing Files (aka unchanged Objects)?

This feature is already present at AWS S3 level once you enable object locking.
![Here is a screenshot](https://i.imgur.com/0tUrdg1.png)

**Comment by maetthu** [comment]
> > Could there be an option to (automatically) extend Object Locks of already existing objects/files while a sync/copy happens?
> > So that it reads the Buckets default retention (for Object Lock) and applies it to all currently existing Files (aka unchanged Objects)?
> 
> This feature is already present at AWS S3 level once you enable object locking.

No, that's not quite what default retention does. When you define a default, **new** objects are configured this way unless you specify a different configuration with request headers in the upload. It doesn't affect existing objects. See https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock-overview.html#object-lock-bucket-config

**Comment by IsSkyfalls** [comment]
A quick search leads to
https://github.com/rclone/rclone/blob/e0d477804b583163b4b5696d24a2ccf8c0bb11e3/backend/s3/s3.go#L6639
. Seems like the headers are not actually added as http headers but parsed
into the request struct. Makes sense when we're using the sdk but not very
flexible.

If you already have a mechanism to add different headers to the request you
probably can also just use some other cli tool to upload the individual
files. Cheers.

**Comment by boogagiga1** [comment]
Thanks for the info, I was trying to avoid that but it is what it is

**Comment by chris081519-crypto** [comment]
Hi everyone,

I've been working on an implementation for S3 Object Lock support and would like to contribute. Before submitting a PR, I wanted to discuss the approach and coordinate with the existing efforts (#7894, #7893, #8866).

### Extended Use Case

Object Lock configurations often have very specific requirements, especially in regulated environments. Each value (mode, retention date, legal hold) should be explicitly configurable - either set to a specific value, copied from the source, or left unchanged.

A common scenario: cloning objects from COMPLIANCE to GOVERNANCE mode while preserving the original retention date is essential for testing in regulated environments.

### Proposed Implementation

Building on the options proposed in #7894, I've added support for a special `copy` value that preserves the source object's setting:

| Option | Values |
|--------|--------|
| `--s3-object-lock-mode` | `GOVERNANCE`, `COMPLIANCE`, or `copy` |
| `--s3-object-lock-retain-until-date` | RFC3339 date, duration (e.g. `365d`, `1y`), or `copy` |
| `--s3-object-lock-legal-hold-status` | `ON`, `OFF`, or `copy` |

The `copy` value means: "read from source object metadata and apply to destination" (requires `--metadata` flag).

**Example - Change mode but keep original retention date:**

```bash
rclone sync source:bucket dest:bucket \
    --metadata \
    --s3-object-lock-mode GOVERNANCE \
    --s3-object-lock-retain-until-date copy
```

### Additional Features I'd Like to Include

Based on the work in #7894, I'd also add:

1. **`--s3-bypass-governance-retention`** - Required for various operations on GOVERNANCE-locked objects, e.g. `sync --delete`

2. **`--s3-bucket-object-lock-enabled`** - Enable Object Lock when creating new buckets with `rclone mkdir`

3. **`--s3-object-lock-set-after-upload`** - For S3 providers that don't support Object Lock headers during PUT (sets retention via separate API calls after upload). This would be an opt-in user flag rather than per-provider quirks, giving users control and avoiding hidden performance impact.

### Questions for Maintainers

1. Should I create a new PR, or would it be better to collaborate with @IsSkyfalls on their existing PR?

2. Is the `copy` value approach acceptable, or would you prefer a different mechanism?

3. Any concerns about the proposed option names?

Happy to adjust the implementation based on feedback. I have a working prototype tested against MinIO.

Thanks!

**Comment by ncw** [comment]
I've merged @chris081519-crypto fix for this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.74
